finesse/hardware/temperature/dp9800.py

```diff
@@ -0,0 +1,238 @@
+        """
+        dev_settings = {
+            "instrument_type": ["TC", "PT"][int(sysflag[0])],
+            "logging_state": ["no logging", "logging active"][int(sysflag[3])],
+            "scanning_state": ["no scan", "autoscan active"][int(sysflag[5])],
+            "audible_state": ["silence", "audible"][int(sysflag[6])],
+            "temperature_unit": ["deg C", "deg F"][int(sysflag[7])],
+        }
+        return dev_settings
+
+    def read(self) -> bytes:
+        """Read temperature data from the DP9800.
+
+        The DP9800 returns a sequence of bytes containing the
+        temperatures measured on each channel, in the format
+
+          STX T SP t1 SP t2 SP t3 SP t4 SP t5 SP t6 SP t7 SP t8 SP t9 ff ETX BCC NUL
+
+        where
+            t1, t2, ..., t9: temperature values in the format %4.2f
+            STX: Start of Text (ASCII 2)
+            ETX: End of Text (ASCII 3)
+            NUL: Null character (ASCII 0)
+            SP: Space (ASCII 32)
+            BCC: Block Check Character
+            ff: System flag in hexadecimal
+
+        For error checking, the BCC is calculated by performing consecutive XOR
+        operations on the message and compared with the BCC received.
+
+        Note a slight peculiarity:
+        The device actually returns 9 temperatures, while the documentation
+        states that there should be 8. The device shows 8 values, corresponding
+        to the values with indices 1 to 8. The first value (at index 0) is
+        therefore ignored.
+
+        Returns:
+            data: the sequence of bytes read from the device
+
+        Raises:
+            DP9800Error: Malformed message received from device
+        """
+        # require at least 4 bytes else checks will fail
+        min_length = 4
+        try:
+            data = self.serial.read_until(b"\x00", size=min_length)
+        except SerialException as e:
+            raise DP9800Error(e)
+
+        if len(data) < min_length:
+            raise DP9800Error("Insufficient data read from device")
+        else:
+            return data
+
+    def request_read(self) -> None:
+        """Write a message to the DP9800 to prepare for a read operation.
+
+        Format:
+
+          EOT T ENQ
+
+        EOT: End of Transmission (ASCII 4)
+        ENQ: Enquiry (ASCII 5)
+
+        Raises:
+            DP9800Error: Error writing to the device
+        """
+        try:
+            self.serial.write(b"\x04T\x05")
+        except Exception as e:
+            raise DP9800Error(e)
+
+    def send_temperatures(self) -> None:
+        """Perform the complete process of reading from the DP9800.
+
+        Writes to the DP9800 requesting a read operation.
+        Reads the raw data from the DP9800.
+        Parses the data and broadcasts the temperatures.
+        """
+        time_now = datetime.now().timestamp()
+        try:
+            self.request_read()
+            data = self.read()
```

Comment:
Generally with timestamps I'd get the time just after reading rather than before, in case the request blocks for a while. (Probably not likely here, but still.)

```suggestion
        try:
            self.request_read()
            data = self.read()
            time_now = datetime.now().timestamp()
```